profiler/src/ProfilerEngine/Datadog.Profiler.Native/ExceptionsProvider.cpp

```diff
@@ -0,0 +1,245 @@
+}
+
+bool ExceptionsProvider::LoadExceptionMetadata()
+{
+    // This is the first observed exception, lazy-load the exception metadata
+    ComPtr<IMetaDataImport2> metadataImportMscorlib;
+
+    INVOKE(
+        _pCorProfilerInfo->GetModuleMetaData(_mscorlibModuleId, CorOpenFlags::ofRead, IID_IMetaDataImport2, reinterpret_cast<IUnknown**>(metadataImportMscorlib.GetAddressOf())))
+
+    mdTypeDef exceptionTypeDef;
+
+    INVOKE(metadataImportMscorlib->FindTypeDefByName(WStr("System.Exception"), mdTokenNil, &exceptionTypeDef))
+
+    ClassID exceptionClassId;
+
+    INVOKE(_pCorProfilerInfo->GetClassFromTokenAndTypeArgs(_mscorlibModuleId, exceptionTypeDef, 0, nullptr, &exceptionClassId));
+
+    ULONG numberOfFields;
+    ULONG classSize;
+
+    INVOKE(_pCorProfilerInfo->GetClassLayout(exceptionClassId, nullptr, 0, &numberOfFields, &classSize));
+
+    const auto fields = std::make_unique<COR_FIELD_OFFSET[]>(numberOfFields);
+
+    INVOKE(_pCorProfilerInfo->GetClassLayout(exceptionClassId, fields.get(), numberOfFields, &numberOfFields, &classSize));
+
+    mdFieldDef messageFieldDef;
+
+    constexpr COR_SIGNATURE signature[2] = {IMAGE_CEE_CS_CALLCONV_FIELD, ELEMENT_TYPE_STRING};
+
+    metadataImportMscorlib->FindField(exceptionTypeDef, WStr("_message"), signature, 2, &messageFieldDef);
+
+    for (ULONG i = 0; i < numberOfFields; i++)
+    {
+        if (fields[i].ridOfField == messageFieldDef)
+        {
+            _messageFieldOffset = fields[i];
+            return true;
+        }
+    }
+
+    return false;
+}
+
+void ExceptionsProvider::DetermineAppDomain(ThreadID threadId, StackSnapshotResultBuffer* const pStackSnapshotResult)
+{
+    // Determine the AppDomain currently running the sampled thread:
+    //
+    // (Note: On Windows, the target thread is still suspended and the AddDomain ID will be correct.
+    // However, on Linux the signal handler that performed the stack walk has finished and the target
+    // thread is making progress again.
+    // So, it is possible that since we walked the stack, the thread's AppDomain changed and the AppDomain ID we
+    // read here does not correspond to the stack sample. In practice we expect this to occur very rarely,
+    // so we accept this for now.
+    // If, however, this is observed frequently enough to present a problem, we will need to move the AppDomain
+    // ID read logic into _pStackFramesCollector->CollectStackSample(). Collectors that suspend the target thread
+    // will be able to read the ID at any time, but non-suspending collectors (e.g. Linux) will need to do it from
+    // within the signal handler. An example for this is the
+    // StackFramesCollectorBase::TryApplyTraceContextDataFromCurrentCollectionThreadToSnapshot() API which exists
+    // to address the same synchronization issue with TraceContextTracking-related data.
+    // There is an additioal complexity with the AppDomain case, because it is likely not safe to call
```

Comment:
Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/2743/commits/329cb7f01524c8dbd502c1e07756e848f524a15c (moved the method to avoid duplication, and fixed the typo)